driver-core/src/main/com/mongodb/connection/WriteCommandProtocol.java

```diff
@@ -81,98 +74,76 @@ public Boolean getBypassDocumentValidation() {
-                }
+        BulkWriteBatchCombiner bulkWriteBatchCombiner = new BulkWriteBatchCombiner(connection.getDescription().getServerAddress(),
+                                                                                          ordered, writeConcern);
+        int batchNum = 1;
+        int currentRangeStartIndex = 0;
+        do {
+            BaseWriteCommandMessage nextMessage = null;
+            batchNum++;
+            if (getLogger().isDebugEnabled()) {
+                getLogger().debug(format("Sending batch %d", batchNum));
+            }
+            ResponseBuffers responseBuffers = connection.sendAndReceive(message);
+            BsonDocument result = createResultDocument(message, responseBuffers);
+            nextMessage = (BaseWriteCommandMessage) message.getEncodingMetadata().getNextMessage();
+            int itemCount = nextMessage != null ? message.getItemCount() - nextMessage.getItemCount() : message.getItemCount();
+            IndexMap indexMap = IndexMap.create(currentRangeStartIndex, itemCount);
+
+            if (getLogger().isDebugEnabled()) {
+                getLogger().debug(format("Received response for batch %d", batchNum));
+            }
 
-                sendSucceededEvent(connection, message, startTimeNanos, result);
+            if (WriteCommandResultHelper.hasError(result)) {
+                MongoBulkWriteException bulkWriteException = getBulkWriteException(getType(), result,
+                        connection.getDescription().getServerAddress());
+                bulkWriteBatchCombiner.addErrorResult(bulkWriteException, indexMap);
+            } else {
+                bulkWriteBatchCombiner.addResult(getBulkWriteResult(getType(), result), indexMap);
+            }
 
-                currentRangeStartIndex += itemCount;
-                message = nextMessage;
-            } while (message != null && !bulkWriteBatchCombiner.shouldStopSendingMoreBatches());
+            currentRangeStartIndex += itemCount;
+            message = nextMessage;
+        } while (message != null && !bulkWriteBatchCombiner.shouldStopSendingMoreBatches());
 
-            return bulkWriteBatchCombiner.getResult();
-        } catch (MongoBulkWriteException e) {
-            throw e;
-        } catch (RuntimeException e) {
-            sendFailedEvent(connection, message, startTimeNanos, e);
-            throw e;
-        }
+        return bulkWriteBatchCombiner.getResult();
     }
 
     @Override
     public void executeAsync(final InternalConnection connection, final SingleResultCallback<BulkWriteResult> callback) {
         executeBatchesAsync(connection, createRequestMessage(getMessageSettings(connection.getDescription())),
-                            new BulkWriteBatchCombiner(connection.getDescription().getServerAddress(), ordered, writeConcern), 0, 0,
+                            new BulkWriteBatchCombiner(connection.getDescription().getServerAddress(), ordered, writeConcern), 1, 0,
```

Comment:
Sort of... The bug is a small logging bug.  Note on line 79 that we're now starting the batchNum at 1 instead of 0, so now we log "Sending batch 1" for the first batch instead of "Sending batch 0".  Made more sense to me that way.